server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -0,0 +1,316 @@
+    IBP_3_1_IV0(3),
+
+    // Support for leader recovery for unclean leader election (KIP-704)
+    IBP_3_2_IV0(4),
+    
+    // KRaft GA
+    IBP_3_3_IV0(5);
+
+    private final Optional<Short> metadataVersion;
+    private final String shortVersion;
+    private final String version;
+
+    MetadataVersion(int metadataVersion) {
+        if (metadataVersion > 0) {
+            this.metadataVersion = Optional.of((short) metadataVersion);
+        } else {
+            this.metadataVersion = Optional.empty();
+        }
+
+        Pattern versionPattern = Pattern.compile("^IBP_([\\d_]+)(?:IV(\\d))?");
+        Matcher matcher = versionPattern.matcher(this.name());
+        if (matcher.find()) {
+            String withoutIV = matcher.group(1);
+            // remove any trailing underscores
+            if (withoutIV.endsWith("_")) {
+                withoutIV = withoutIV.substring(0, withoutIV.length() - 1);
+            }
+            shortVersion = withoutIV.replace("_", ".");
+
+            if (matcher.group(2) != null) { // versions less than IBP_0_10_0_IV0 do not have IVs
+                version = String.format("%s-IV%s", shortVersion, matcher.group(2));
+            } else {
+                version = shortVersion;
+            }
+        } else {
+            throw new IllegalArgumentException("Metadata version: " + this.name() + " does not fit "
+                + "the accepted pattern.");
+        }
+    }
+
+    public Optional<Short> metadataVersion() {
+        return metadataVersion;
+    }
+
+    public boolean isSaslInterBrokerHandshakeRequestEnabled() {
+        return this.isAtLeast(IBP_0_10_0_IV1);
+    }
+
+    public boolean isOffsetForLeaderEpochSupported() {
+        return this.isAtLeast(IBP_0_11_0_IV2);
+    }
+
+    public boolean isFeatureVersioningSupported() {
+        return this.isAtLeast(IBP_2_7_IV0);
+    }
+
+    public boolean isTruncationOnFetchSupported() {
+        return this.isAtLeast(IBP_2_7_IV1);
+    }
+
+    public boolean isAlterIsrSupported() {
+        return this.isAtLeast(IBP_2_7_IV2);
+    }
+
+    public boolean isTopicIdsSupported() {
+        return this.isAtLeast(IBP_2_8_IV0);
+    }
+
+    public boolean isAllocateProducerIdsSupported() {
+        return this.isAtLeast(IBP_3_0_IV0);
+    }
+
+
+    public RecordVersion recordVersion() {
+        if (this.isLessThan(IBP_0_10_0_IV0)) {
+            // IBPs less than IBP_0_10_0_IV0 use Record Version V0
+            return RecordVersion.V0;
+        } else if (this.isLessThan(IBP_0_11_0_IV0)) {
+            // IBPs >= IBP_0_10_0_IV0 and less than IBP_0_11_0_IV0 use V1
+            return RecordVersion.V1;
+        } else {
+            // all greater IBPs use V2
+            return RecordVersion.V2;
+        }
+    }
+
+    private static final Map<String, MetadataVersion> IBP_VERSIONS;
+    static {
+        {
+            IBP_VERSIONS = new HashMap<>();
+            Pattern versionPattern = Pattern.compile("^IBP_([\\d_]+)(?:IV(\\d))?");
```

Comment:
The old code used case objects for the different IBP versions and populated the equivalent map using hardcoded `shortVersion` and `version` fields. This seems unnecessary since the same information can be parsed out from the enum constant names.

Reply:
Parsing is generally less readable/maintainable though.

Reply:
I guess it's a tradeoff of concise/clean code and readability, especially the IBP versions do not share the same format (some have IVs, some have 2 numerals before the IV, some have 3 before the IV). 
I think my preference would be to just better document/clean up this parsing logic to make it very clear what it is doing, but as an alternative to this parsing, I could add the `shortVersion` and `subVersion` as fields to the enum constants? 
```
public enum MetadataVersion {
    IBP_0_8_0(-1, "0.8.0", ""),
    ...
    IBP_0_10_0_IV0(-1, "0.10.0", "IV0"),
    ...
    ...
    
    private final Optional<Short> metadataVersion;
    private final String shortVersion;
    private final Optional<String> ivVersion;
    
    MetadataVersion(int metadataVersion, String shortVersion, String ivVersion) {
    ...
    } 
```
cc @ijuma @mumrah

Reply:
Conciseness is good when it aids in readability. If you have to add documentation to explain the concise code, that's usually not a good thing. The version where you pass the two parameters to the enum constructor seems a lot cleaner to me.

Reply:
@ijuma I've updated the MetadataVersion enum as described above!

Reply:
Yeah, we'd better not change this logic since otherwise will bring some unexpected things, for example, we can still parse a misconfigured ibp value "2.1.1" to "2.1-IV2" in the past, maybe we should also keep parsing some of these misconfigured value.

Reply:
^ the same behavior is preserved with this refactor (there's an existing test case which tests the equivalent behavior -> `assertEquals(IBP_1_0_IV0, MetadataVersion.fromVersionString("1.0.1"));`

Reply:
Can you add a comment about this to the parsing function? I honestly consider this a bug but if we want to be bug-compatible, we should explicitly note that that's what we're doing (since otherwise someone might "fix" it)